File: src/sender/crc32.cpp
#include <iostream>
#include <string>
#include "crc32.h"

int mostSignificantBitPosition(const std::string& bitString) {
    for (int i = 0; i < bitString.size(); ++i) {
        if (bitString[i] == '1') {
            return i;
        }
    }
    return -1;
}

void XOR(std::string& data, const std::string& polynomial, const int pos){
    for(int i = 0; i<polynomial.size(); i++){
        if(data[i + pos] == polynomial[i]){
            data[i+pos] = '0';
        }
        else{
            data[i+pos] = '1';
        }
    }
}

std::string crc32(const std::string& data, const std::string& polynomial, const int data_len) {
    std::string result = data;
    int pos = 0;
    while((data.size() - pos) >= polynomial.size() - 1){
        XOR(result, polynomial, pos);
        pos = mostSignificantBitPosition(result);
    }
    return result.substr(data.size() - (polynomial.size() - 1));
}

File: src/receiver/crc32.py
def XOR(data, poly, pos):
    data_list = list(data)
    for x in range(len(poly)):
        if data_list[pos + x] == poly[x]:
            data_list[pos + x] = '0'
        else:
            data_list[pos+x] = '1'
    return ''.join(data_list)
    
def mostSignificantBit(data):
    for x in range(len(data)):
        if data[x] == '1':
            return x
    else:
        return -1
        
def CRC32(data, poly):
    pos = 0
    while ((len(data) - pos) >= len(poly)):
        data = XOR(data, poly, pos)
        pos = mostSignificantBit(data)
        if pos == -1: break
    checksum = data[-(len(poly)-1):]
    f = checksum.find('1')
    return f == -1, data
    

if __name__ == "__main__"   :
    received = "1101001000000000000000000000000010110"
    poly = "111100000000000000000000000000001"
    flag, result = CRC32(received, poly)
    if not flag:
        print(f"Se encontró un error. La cadena resultante es: {result}")
    else: 
        print(f"Mensaje recibido correctamente. Cadena resultante: {result}")
    